Add kubelet metrics for ephemeral containers

verb opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/sig node
/priority important-soon

#### What this PR does / why we need it:

This adds alpha metrics to the kubelet that track pods and containers under management and counters for container creation. This is part of the Ephemeral Containers (kubernetes/enhancements#277) PRR.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #97974

#### Special notes for your reviewer:

As part of code cleanup this slightly changes the format of log messages of particular container types. This isn't necessary and can be reverted, but using the same label in log messages in metrics seems cleaner and more supportable.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://git.k8s.io/enhancements/keps/sig-node/277-ephemeral-containers
```

verb committed:
Add kubelet managed pod metrics

verb commented:
/retest

verb commented:
I was going to ask @ehashman for advice on these metrics, but I just found [sig-instrumentation docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md) that make me think that I've got it approximately right.

I'll go ahead finish up my TODO to add tests, remove the WIP hold and then submit it for review.

ehashman commented:
Let me know when this is ready for review and I'll TAL.

verb commented:
/retest

verb commented:
Looks like timeouts that are unrelated to this change

/retest

kikisdeliveryservice commented:
/triage accepted

verb commented:
Thanks @ehashman! It looks like metrics aren't usually tested, so this is ready for your review.

/assign @ehashman

ehashman commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -840,14 +847,14 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
 	// containers cannot be specified on pod creation.
 	if utilfeature.DefaultFeatureGate.Enabled(features.EphemeralContainers) {
 		for _, idx := range podContainerChanges.EphemeralContainersToStart {
-			start("ephemeral container", ephemeralContainerStartSpec(&pod.Spec.EphemeralContainers[idx]))
+			start(metrics.EphemeralContainerLabel, ephemeralContainerStartSpec(&pod.Spec.EphemeralContainers[idx]))
```
I think we need an `ACTION REQUIRED:` release note since these labels are changing.

verb replied:
Sure, happy to do that. We could also keep the old format for the low, low price of more code. What do you think?

verb replied:
Since we introduced snake case for the metric labels I reverted this change. Log messages will remain unchanged.

ehashman commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -379,6 +392,64 @@ var (
+		},
+		[]string{"container_type", "code"},
+	)
+	// ManagedPods is a gauge that tracks how many pods are managed by this kubelet
+	ManagedPods = metrics.NewGauge(
```
The name here is a bit confusing to me; why "managed" pods? Is this to differentiate from static/mirror pods?

verb replied:
I was thinking "pods managed by this kubelet", but naming is hard and I'm happy to change it.

ehashman commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -379,6 +392,64 @@ var (
 		},
 		[]string{"container_state"},
 	)
+	// StartedPodsTotal is a counter that tracks pod sandbox creation operations
```
/cc @lilic 

these metrics are currently provided by KSM but I think it would also be useful to be able to compare them to the kubelet-level view

ehashman reviewed (commented):
We do have the ability to do some metric testing via component-base fwiw, although I don't know if we do this in any components currently: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/collector_test.go#L93

verb reviewed (commented):
@ehashman sorry for the radio silence here. FYI, we've delayed beta so there's no rush to try to merge this by freeze.

lilic commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -79,6 +79,19 @@ const (
+	// Metrics to keep track of objects under management
+	ManagedPodsKey                  = "managed_pods"
+	ManagedContainersKey            = "managed_containers"
+	StartedPodsTotalKey             = "started_pods_total"
```
We have kube_pod_start_time and kube_pod_info in kube-state-metrics, which can be summed into the total started pods. Note we have actual drill down of pod name. I guess if folks don't use KSM this can be useful and its not too much cardinality.

https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md

lilic commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -79,6 +79,19 @@ const (
+	ManagedContainersKey            = "managed_containers"
+	StartedPodsTotalKey             = "started_pods_total"
+	StartedPodsErrorsTotalKey       = "started_pods_errors_total"
+	StartedContainersTotalKey       = "started_containers_total"
```
Same as above we have containers and init containers. Do you plan on adding init containers metrics as well?

verb replied:
Yes, the type of container is stored using the label `container_type` rather than a separate metric.

lilic commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -79,6 +79,19 @@ const (
+
+	// Labels used in metric fields
+	ContainerLabel          = "container"
+	InitContainerLabel      = "initContainer"
```
Nit: This should be `init_container` instead, same for `ephemeralContainer` below -> `ephemeral_container`.

verb replied:
Ok, updated these to `init_container` and `ephemeral_container`.

I think I used the wrong terminology in the comment (I'll go back and fix this), so just to make sure I didn't mislead: these are the values used in the label named `container_type`. I arbitrarily chose the names used in the API (`container`, `initContainer`, `ephemeralContainer`)  because I didn't know the convention for label values.

So previously it was:

```
# HELP kubelet_started_containers_total [ALPHA] Cumulative number of containers started
# TYPE kubelet_started_containers_total counter
kubelet_started_containers_total{container_type="container"} 5
kubelet_started_containers_total{container_type="initContainer"} 1
```

and now it's:
```
# HELP kubelet_started_containers_total [ALPHA] Cumulative number of containers started
# TYPE kubelet_started_containers_total counter
kubelet_started_containers_total{container_type="container"} 5
kubelet_started_containers_total{container_type="init_container"} 1
```

lilic replied:
Looks good to me, thank you!

lilic reviewed (commented):
Thanks @ehashman for the ping!

Nothing against instrumentation in the kubelet, just pointed out the metrics we have in kube-state-metrics already. It might be helpful to have a few labels to be able to aggregate on these metrics, we tend to use namespace, host/node and container id so we can group and alert on specific errors coming from specific kubelet.

verb commented:
@ehashman @lilic Thanks for your reviews on this. The remaining open questions are:

1. Do we want to surface these metrics in the kubelet even though there's overlap with KSM? (Sounds like the general feeling is yes.)
2. What should we name the gauge metric that indicates the number of running pods and containers.

Here's an example of the metrics added by this PR in its current state.

Counters:
```
# HELP kubelet_started_containers_errors_total [ALPHA] Cumulative number of errors when starting containers
# TYPE kubelet_started_containers_errors_total counter
kubelet_started_containers_errors_total{code="ErrImagePull",container_type="ephemeral_container"} 2
kubelet_started_containers_errors_total{code="ImagePullBackOff",container_type="ephemeral_container"} 1
# HELP kubelet_started_containers_total [ALPHA] Cumulative number of containers started
# TYPE kubelet_started_containers_total counter
kubelet_started_containers_total{container_type="container"} 10
kubelet_started_containers_total{container_type="ephemeral_container"} 4
kubelet_started_containers_total{container_type="init_container"} 1
# HELP kubelet_started_pods_total [ALPHA] Cumulative number of pods started
# TYPE kubelet_started_pods_total counter
kubelet_started_pods_total 2
```

Gauges:
```
# HELP kubelet_managed_containers [ALPHA] Number of containers managed by this kubelet
# TYPE kubelet_managed_containers gauge
kubelet_managed_containers{container_type="container"} 3
kubelet_managed_containers{container_type="ephemeral_container"} 2
kubelet_managed_containers{container_type="init_container"} 1
# HELP kubelet_managed_pods [ALPHA] Number of pods managed by this kubelet
# TYPE kubelet_managed_pods gauge
kubelet_managed_pods 2
```

I named the gauges `kubelet_managed_` thinking "managed by these kubelet", but maybe this is confusing. @ehashman do you have a suggestion for a better name?

ehashman reviewed (commented):
For 1, I have no issue with there being some overlap given that the cardinality is so limited.

For 2, I still have some concerns. Right now, this PR introduces new metrics such that all the following exist:

```
		legacyregistry.MustRegister(RunningContainerCount)
		legacyregistry.MustRegister(RunningPodCount)
		legacyregistry.MustRegister(ManagedPods)
		legacyregistry.MustRegister(ManagedContainers)
```

It is very unclear to me as an end user what the difference between "running" and "managed" pods is. "Running" is intuitive enough (number of pods currently in phase "Running"), but "managed" has not been defined. If it means "all phases that are not Terminated", I think we should a) say that explicitly, and explicitly call out the phases included in the help text b) call them active pods, not managed.

verb commented:
@ehashman The real reason there are two similar metrics is a side effect of how they're implemented. The thing I want to measure isn't available to the existing metrics, so either we measure in two places or plumb more metadata down to the metric. I started with the former, but we may decide to change to the latter. It has other benefits, but it's a bigger change.

The `running_*` metrics measure the internal kubelet representation of pods and containers, measured in the PLEG. The `managed_*` metrics measure `v1.Pod` objects tracked by the kubelet's pod manager (hence "managed").

* Completed pods appear in `managed_pods` but not in `running_pods`, which makes sense.
* `running_pods` is actually measuring running sandboxes
* Pods in `*BackOff` appear in both `managed_pods` and `running_pods` because the sandbox is running, even if the "pod" isn't.
* `running_containers` aren't necessarily running. They're only running if the `container_state` label is `running` rather than `created` or `exited`.

`running_pods` and `running_containers` can be misleading if one expects them to be related to Kubernetes objects. For some reason my test cluster is showing running containers even though i'm not sure what they are:

```
% kubectl get -A po ; docker ps; kubelet-metrics |  egrep 'kubelet_(running|managed)_containers'
No resources found
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
# HELP kubelet_managed_containers [ALPHA] Number of containers managed by this kubelet
# TYPE kubelet_managed_containers gauge
kubelet_managed_containers{container_type="container"} 0
kubelet_managed_containers{container_type="ephemeral_container"} 0
kubelet_managed_containers{container_type="init_container"} 0
# HELP kubelet_running_containers [ALPHA] Number of containers currently running
# TYPE kubelet_running_containers gauge
kubelet_running_containers{container_state="created"} 1
kubelet_running_containers{container_state="exited"} 1
kubelet_running_containers{container_state="running"} 1
```

I intended "managed" to mean "pods managed by this kubelet". It's "all pods even if they're terminated or not started yet"

Alternatively, I could plumb the `container_type` metadata down to the kubelet's intermal representation and add this information into the existing `running_*` metrics. This would also allow us to set `container_type` labels in the runtimes, which containerd recently expressed interest in having.

verb commented:
So, we can either:

1. Have 2 sets of metrics: one that reflect Kubernetes API objects (v1.Pod, v1.Container) that this kubelet is tracking and a second that reflect the kubelet's internal representation of containers and sandboxes.
2. Modify the kubelet's internal representation of containers to have container type information.

I prefer the first because both sets of metrics are interesting, and they're interesting to compare to each other. If we choose this then I will update the descriptions of both metrics so that the difference is clear, and we can change "managed" to something more descriptive.

The second requires changing how the kubelet works in order to add monitoring, which seems backwards, but it could have other benefits.

@ehashman Do you agree? Should I try to improve the metric descriptions and see how it looks?

dchen1107 commented:
cc/ @dashpole can you please take a look too? Thanks!

dashpole commented:
> I prefer the first because both sets of metrics are interesting, and they're interesting to compare to each other.

Some of the differences are interesting, such as if a pod sandbox is failing to be created, but most of the differences you mention are probably just bugs we should fix.  Under "normal" conditions, i'd expect the two to match (sandboxes should match pods, etc.).  I'm not sure if the differences between the two are interesting from the point of view of the ephemeral containers feature, though?  It seems like option 2 would be needed even if we decide to do option 1, since you want to compare the two

verb commented:
@dashpole Thanks for the guidance here. No, I don't need to compare these two. For ephemeral containers I really only want to answer two questions:

1. How many ephemeral containers are running on this node?
2. How many ephemeral containers has this node started? were there errors?

The kubelet discards container type information, so the pod manager was the most convenient place to track the first. It's still unclear on what we want to measure. I plan on bringing it up at tonight's sig-node meetup.

Random-Liu commented:
I always treated "pod manager" just a local cache of the pods in apiserver, thus it contains the "desired state", and no much different with counting pods from the apiserver.

However, I do agree with the point brought in the meeting that, having the pods in the apiserver doesn't necessarily mean that kubelet has seen it:
1) There might be a bug that kubelet doesn't see a pod;
2) There might be a delay before kubelet sees the pod.

So in theory, the "pod manager" does contain some fact about the actual state - "whether kubelet has seen the pods".

However, it is not quite clear to me how useful it is to introduce metrics counting for pods in the pod manager.
For case 1) debug, I think the pod status is a better indication of the actual issue, than comparing the pod number between apiserver and managed pods.
For case 2), if the delay is huge, the pod events and status can be used to debug the issue; if the delay is minimal, the metrics is almost the same with the counting on the apiserver side.

Given so, I'm not sure adding metrics for managed_pods actually brings enough value.

The started containers/pods do make sense to me, because it exposes more useful actual state from kubelet which can be useful for trouble shooting.

ehashman commented:
Waiting on changes per the above from @Random-Liu, thanks @verb for all the work you've done here :)

verb commented:
Ah, thanks for pinging this. If I understand correctly, both @Random-Liu and @dchen1107 are in favor of changing the "managed {containers,pods}" gauge to be narrowly scoped for ephemeral containers, and @derekwaynecarr thought the additional information for all types would be useful.

I would want metrics at every stage of an automation system. I agree that logs are better for debugging kubelet problems, but these metrics are for observing cluster state rather than debugging problems with the kubelet. Surfacing information on the kubelet's internal state, even if it is theoretically available from the API server, is useful. The API server may be using different feature flags or perhaps it's unreachable.

I think @dchen1107's point was that solving this problem deserves a full design rather than a hasty solution to meet the PRR requirements for a single feature, so we should add a metric that's specific to that PRR. On the other hand, it seems strange to me to provide this information for ephemeral containers and withhold it for other types of containers.

I can think of 3 different ways forward:
1. Add `container_type` to the existing `running_containers` metrics. This is more monitoring driving design, so let's avoid it.
2. All of these metrics are marked `alpha`. We could continue to add alpha metrics and that way, when someone has time to design the kubelet metrics API, there's a ready list of metrics that people actually want with no commitment to support them indefinitely.
3. Add a `managed_ephemeral_containers` metric, as suggested by @Random-Liu.

I like the idea of there being some mechanism of iterating on unstable metrics (as in the second), but I don't have time to work on that right now, so I'll just plan on the third option.

verb committed:
Remove ManagedPod,ManagedContainer metrics

This replaces the generic ManagedPod and ManagedContainer kubelet
metrics with a gauge to track only ephemeral container usage.

verb commented:
/retest

verb commented:
@ehashman PTAL! Thanks :smile:

ehashman commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -431,6 +445,54 @@ var (
 		[]string{"container_state"},
 	)
+	// StartedPodsTotal is a counter that tracks pod sandbox creation operations
+	StartedPodsTotal = metrics.NewCounter(
```
Note to self: this one is `NewCounter` but the rest of these new metrics are `NewCounterVec` because this one doesn't have any additional labels (e.g. container type, error).

ehashman reviewed (commented):
/lgtm

verb commented:
/assign @klueska

verb commented:
Hi @ehashman, I see you added this to the "Needs Approver" board. Is there a process for assigning the Approver or should I seek one out? Thanks!

verb commented:
/assign @dchen1107

dchen1107 commented:
/lgtm 
/approve

Thanks for narrow down the scope to the ephemeral containers, and renamed some of metrics per reviewers' and SIG Node's suggests. Also two reviewers from SIG instruments are ok with the small overlap even before the narrow down. 

> However, it is not quite clear to me how useful it is to introduce metrics counting pods in the pod manager.
For case 1), if it happens it is a bug we should troubleshoot, I think the pod events/status and kubelet log is a more useful indication of the actual issue, than comparing the pod number between apiserver and managed pods.
For case 2), if the delay is huge, it is also a bug we should troubleshoot, and the pod events/status and kubelet log can be used to debug the issue, just counting the number may not be that useful; if the delay is minimal, the metrics is almost the same with the counting on the apiserver side.

Discussed this at SIG Node a while back, and here is my take on this very topic: Logging and Metrics are serving the different purpose for observability and debuggability. Metrics are usually measured a time series, and can be optimized for the storage and longer retention of data, and easily build dashboards to reflect historical trends and be aggregated into daily or weekly frequency. This is why normally admins / SREs prefer to use metrics to detect the abnormal states, and the developer looking into the logs for the troubleshooting. 

Back to this particular PR, I have to admit that the current metrics being introduced here might be the best effort. In a real environment, kubelet might be restarted due to different reasons. Once the kubelet is restarted, the count of the container starts with error would be lost, even the current running containers & pods can be re-generated through the current state. Just want to point it out this.

ehashman commented:
/milestone v1.22
